pop-api/examples/fungibles/tests.rs

```diff
@@ -0,0 +1,831 @@
+			admin: account_id_from_slice(contract.as_ref()),
+		}
+		.encode()
+		.as_slice()
+	);
+}
+
+#[drink::test(sandbox = Sandbox)]
+fn existing_constructor_works(mut session: Session) {
+	let _ = env_logger::try_init();
+	// Fails to deploy contract with a non-existing token ID.
+	assert_eq!(
+		deploy(&mut session, CONTRACT.clone(), "existing", vec![TOKEN.to_string()]),
+		Err(PSP22Error::Custom(String::from("Unknown")))
+	);
+
+	// Successfully deploy contract with an existing token ID.
+	let actor = session.get_actor();
+	assert_ok!(session.sandbox().create(&TOKEN, &actor, MIN_BALANCE));
+	assert_ok!(deploy(&mut session, CONTRACT.clone(), "existing", vec![TOKEN.to_string()]));
+	assert!(session.sandbox().asset_exists(&TOKEN));
+}
+
+/// 1. PSP-22 Interface:
+/// - total_supply
+/// - balance_of
+/// - allowance
+/// - transfer
+/// - transfer_from
+/// - approve
+/// - increase_allowance
+/// - decrease_allowance
+
+#[drink::test(sandbox = Sandbox)]
+fn total_supply_works(mut session: Session) {
+	let _ = env_logger::try_init();
+	// Deploy a new contract.
+	assert_ok!(deploy(
+		&mut session,
+		CONTRACT.clone(),
+		"new",
+		vec![TOKEN.to_string(), MIN_BALANCE.to_string()]
+	));
+
+	// No tokens in circulation.
+	assert_eq!(total_supply(&mut session), 0);
+	assert_eq!(total_supply(&mut session), session.sandbox().total_supply(&TOKEN));
+
+	// Tokens in circulation.
+	assert_ok!(session.sandbox().mint_into(&TOKEN, &ALICE, AMOUNT));
+	assert_eq!(total_supply(&mut session), AMOUNT);
+	assert_eq!(total_supply(&mut session), session.sandbox().total_supply(&TOKEN));
+}
+
+#[drink::test(sandbox = Sandbox)]
+fn balance_of_works(mut session: Session) {
+	let _ = env_logger::try_init();
+	// Deploy a new contract.
+	assert_ok!(deploy(
+		&mut session,
+		CONTRACT.clone(),
+		"new",
+		vec![TOKEN.to_string(), MIN_BALANCE.to_string()]
+	));
+
+	// No tokens in circulation.
+	assert_eq!(balance_of(&mut session, ALICE), 0);
+	assert_eq!(balance_of(&mut session, ALICE), session.sandbox().balance_of(&TOKEN, &ALICE));
+
+	// Tokens in circulation.
+	assert_ok!(session.sandbox().mint_into(&TOKEN, &ALICE, AMOUNT));
+	assert_eq!(balance_of(&mut session, ALICE), AMOUNT);
+	assert_eq!(balance_of(&mut session, ALICE), session.sandbox().balance_of(&TOKEN, &ALICE));
+}
+
+#[drink::test(sandbox = Sandbox)]
+fn allowance_works(mut session: Session) {
+	let _ = env_logger::try_init();
+	// Deploy a new contract.
+	let contract = deploy(
+		&mut session,
+		CONTRACT.clone(),
+		"new",
+		vec![TOKEN.to_string(), MIN_BALANCE.to_string()],
+	)
+	.unwrap();
+
+	// No tokens in circulation.
+	assert_eq!(allowance(&mut session, contract.clone(), ALICE), 0);
+
+	// Tokens in circulation.
+	assert_ok!(session.sandbox().approve(&TOKEN, &contract.clone(), &ALICE, AMOUNT / 2));
+	assert_eq!(allowance(&mut session, contract, ALICE), AMOUNT / 2);
+}
+
+#[drink::test(sandbox = Sandbox)]
+fn transfer_works(mut session: Session) {
+	let _ = env_logger::try_init();
```

Comment:
Taking inspiration from new_test().execute_with(|| { }), an improvement might be to define something similar in pop-drink which takes the session and contract details as parameters and then has a function for executing a closure, with the session and contract address as the closure parameters.

Reply:
Something for later as I don't feel this necessary, I'm sorry but I have to draw a line somewhere. Created an issue in pop drink repo.

Reply:
I will resolve this comment. There was an issue created: https://github.com/r0gue-io/pop-drink/issues/9